### PR TITLE
enable customizing cursor navigation behaviour

### PR DESCRIPTION
**Pull request by rburgst**
- with this change the container can inject a custom strategy
   that allows customization of possible selectable cells
- one example use case is to allow skipping over
  read-only fields

### CODE REVIEW

**Comment by rburgst** [comment]
anyone home?

File: src/DataSheet.js [hunk 1]
```diff
@@ -349,6 +350,38 @@ export default class DataSheet extends PureComponent {
         }
         return false
       }
+
+      if (this.props.allowNav) {
```
Comment (nadbm) [inline src/DataSheet.js]: "allowNav" seems like a boolean here. 
maybe isCellNavigable(cell) would be a better fit.

Comment (rburgst) [reply src/DataSheet.js]: done

**Comment by rburgst** [comment]
ping, I have implemented the requested changes, can you re-check?

**Comment by nadbm** [comment]
I'm trying this locally and this is not exactly in working. 

in docs/examples/BasicSheet I added the following to skip over row 3: 
`isCellNavigable={(cell,i,j,jump) => return i === 2}`
This unfortunately causes an infinite loop. 

Im sorry the handleNavigateMethod got so complex, it's hard to add any changes. I will push a PR soon to refactor that. That might help you debug your issue.

**Comment by rburgst** [comment]
> I'm trying this locally and this is not exactly in working.
> 
> in docs/examples/BasicSheet I added the following to skip over row 3:
> `isCellNavigable={(cell,i,j,jump) => return i === 2}`
> This unfortunately causes an infinite loop.
> 
Can you let me  know what the selected cell was and what key you used?

**Comment by rburgst** [comment]
or maybe even better reproduce it in the unit test? I have already created a few.

**Comment by nadbm** [comment]
```js
  describe('DataSheet component with restricted cell movement', () => {
    let data = []
    let component = null
    let wrapper = null
    let customWrapper = null
    let selected = null
    jsdom()

    beforeEach(() => {
      data = [
        [{
          className: 'test1',
          data: 4,
          overflow: 'clip'
        }, {
          className: 'test2',
          data: 2,
          key: 'custom_key'
        }, {
          className: 'test3',
          data: 3,
        }, {
          className: 'test4',
          data: 4,
        }],
        [{
          className: 'test5',
          data: 0,
          width: '25%'
        }, {
          className: 'test6',
          data: 5,
          width: 100
        }, {
          className: 'test7',
          data: 5,
          width: 100
        }, {
          className: 'test8',
          data: 3,
        }]
      ]
      component = <DataSheet
        keyFn={(i) => 'custom_key_' + i}
        className={'test'}
        overflow='nowrap'
        data={data}
        valueRenderer={(cell) => cell.data}
        // only allow navigation to the middle 2 columns
        isCellNavigable={(row, col) => row === 2 }
        onChange={(cell, i, j, value) => data[i][j].data = value}
      />
      wrapper = mount(component)
    })
    afterEach(() => {
      wrapper.instance().removeAllListeners()
      if (customWrapper) {
        if ('removeAllListeners' in customWrapper.instance()) {
          customWrapper.instance().removeAllListeners()
        }
        customWrapper = null
      }
    })

    it('navigate with tab over restricted cells', () => {
      // click into row0, col1
      wrapper.find('td').at(1).simulate('mouseDown')
      expect(wrapper.state('start')).toEqual({i: 0, j: 1})
      // first tab gets us to row0, col2
      triggerKeyDownEvent(wrapper, DOWN_KEY)
      expect(wrapper.state('start')).toEqual({i: 1, j: 1})      
    })

  })
```

**Comment by rburgst** [comment]
Thanks a lot for the test case, I was too focused on horizontal movement to have missed the vertical movement and its issues. However, your test case can never pass correctly since you have only 2 rows but restrict movement to row 3, therefore, none of the existing rows are navigable.
I fixed the infinite loop and adapted the test case accordingly.

**Comment by rburgst** [comment]
note that I rebased on the current master

**Pull request merged**